bootloader/src/pyi_utils.c

```diff
@@ -1028,159 +1057,279 @@ pyi_utils_create_child(const char *thisfile, const ARCHIVE_STATUS* status,
+                    VS("LOADER [AppleEvent ARGV_EMU]: arg[%d] = %s\n", (int)argc_pyi, buf);
+                    tmp_str = strdup(buf);
+                    tmp_argv = (char **)realloc(argv_pyi, (argc_pyi + 2) * sizeof(char *));
+                    if (!tmp_argv || !tmp_str) {
+                        /* Out of memory. Extremely unlikely -- not clear what to do here.
+                         * Attempt to silently continue. */
+                        VS("LOADER [AppleEvent ARGV_EMU]: allocation for arg[%d] failed: %s\n",
+                           argc_pyi, strerror(errno));
+                        free(tmp_argv); /* free of NULL ok */
+                        free(tmp_str);
+                        continue;
+                    }
+                    argv_pyi = tmp_argv;
+                    argv_pyi[argc_pyi++] = tmp_str;
+                    argv_pyi[argc_pyi] = NULL;
+                    VS("LOADER [AppleEvent ARGV_EMU]: argv entry appended.\n");
+                }
+            }
 
-     argv_pyi = (char**)realloc(argv_pyi,(argc_pyi+2)*sizeof(char*));
-     argv_pyi[argc_pyi++] = strdup(buffer);
-     argv_pyi[argc_pyi] = NULL;
+            err = AEDisposeDesc(&docList);
 
-     VS("LOADER [ARGV_EMU]: argv entry appended.");
+            return err;
 
-     free(buffer);
-   }
+        } else {
+            /* The child process exists.. so we forward events to it */
 
-  err = AEDisposeDesc(&docList);
+            ProcessSerialNumber psn;
+            AppleEvent evtCopy;
+            AEAddressDesc target;
+            OSStatus err = noErr;
+            AEEventClass evtClass;
+            AEEventID evtID;
 
+            if (apple_event_is_open_uri) {
+                evtClass = kInternetEventClass;
+                evtID = kAEGetURL;
+            } else {
+                /* open doc */
+                evtClass = kCoreEventClass;
+                evtID = kAEOpenDocuments;
+            }
 
-  return (err);
+            VS("LOADER [AppleEvent EVT_FWD]: Will forward %s event to child pid %ld...\n", descStr, (long)child_pid);
+
+            err = GetProcessForPID(child_pid, &psn);
+            if (err != noErr) return (OSErr)err;
+            VS("LOADER [AppleEvent EVT_FWD]: Creating target desc.\n");
+            /* re-target the event to our child process: creating a descriptor and event object for it */
+            err = AECreateDesc(typeProcessSerialNumber, &psn, sizeof(psn), &target);
+            if (err != noErr) return (OSErr)err;
+            VS("LOADER [AppleEvent EVT_FWD]: Creating dupe event.\n");
+            err = AECreateAppleEvent(evtClass, evtID, &target, kAutoGenerateReturnID, kAnyTransactionID, &evtCopy);
+            if (err != noErr) goto cleanup2;
+
+            { /* <-- These curly braces aren't a typo. We are creating a nested scope here for below vars. */
+                char buf[64*1024]; /* 64 KiB buffer should be enough for drag/drops */
```

Comment:
Out of curiosity, do you have numbers for the amount of paths that can be drag'n droped with this value? I guess with 10,000 files (or even more, I have use cases :) it would not work?

Reply:
Hmm you know what? I really should characterize this.  I haven't tried and have no idea what would happen.  

The app wouldn't crash but worst-case it would just reject the huge mass of files (because buffer is too small), and ignore them.

I will characterize this further.

Reply:
FTR I have use case when users would drag'n drop 50,000 or even 70,000 files at once (depending of what Finder allow to select without freezing).

Reply:
It's something I can see myself doing too in some apps -- definitely worth exploring further as long as we are here.

Reply:
OK, I believe this has been resolved.  There are 2 cases:

1. At app startup it loops through all the items in the AppleEvent, using a 64KiB static buffer for each (due to the way the code is written I opted to use a static buffer here).  This 64KiB per item should be enough. There can be potentially infinite items, depending on how much memory the user has. Should be fine.

2. As the app is running it now uses a dynamically-allocated buffer for the event itself -- which it then copies, rewrites destination, and forwards (sort of like doing NAT on IPv4, heh).  So this should handle anything macOS throws at it.

Reply:
Great news :)